File: client/legacy-gabriel-client/src/edu/cmu/cs/gabriel/token/TokenController.java
package edu.cmu.cs.gabriel.token;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.concurrent.ConcurrentHashMap;

import android.os.Handler;
import android.os.Message;
import android.util.Log;
import edu.cmu.cs.gabriel.Const;
import edu.cmu.cs.gabriel.network.NetworkProtocol;

public class TokenController {
    private static final String LOG_TAG = "TokenController";

    // the number of tokens remained
    private int currentToken = 0;
    
    // information about all sent packets, the key is the frameID and the value documents relevant timestamps
    private ConcurrentHashMap<Long, SentPacketInfo> sentPackets = new ConcurrentHashMap<Long, SentPacketInfo>();
    
    private Object tokenLock = new Object();
    
    private FileWriter fileWriter = null;

    // timestamp when the last ACK was received
    private long prevRecvFrameID = 0;

    public TokenController(int tokenSize, File resultSavingPath) {
        this.currentToken = tokenSize;
        if (Const.IS_EXPERIMENT) {
            try {
                fileWriter = new FileWriter(resultSavingPath);
                fileWriter.write("FrameID\tEngineID\tStartTime\tCompressedTime\tRecvTime\tDoneTime\tStatus\n");
            } catch (IOException e) {
                Log.e(LOG_TAG, "Result file cannot be properly opened", e);
            }
        }
    }

    public Handler tokenHandler = new Handler() {

        public void handleMessage(Message msg) {
            if (msg.what == NetworkProtocol.NETWORK_RET_SYNC) {
                try {
                    if (Const.IS_EXPERIMENT){
                        String log = (String) msg.obj;
                        fileWriter.write(log);
                    }
                } catch (IOException e) {}
            }
            if (msg.what == NetworkProtocol.NETWORK_RET_TOKEN) {
                ReceivedPacketInfo receivedPacket = (ReceivedPacketInfo) msg.obj;
                long recvFrameID = receivedPacket.frameID;
                String recvEngineID = receivedPacket.engineID;
                
                // increase appropriate amount of tokens
                long increaseCount = 0;
                for (long frameID = prevRecvFrameID + 1; frameID < recvFrameID; frameID++) {
                    SentPacketInfo sentPacket = null;
                    if (Const.IS_EXPERIMENT) {
                        // Do not remove since we need to measure latency even for the late response
                        sentPacket = sentPackets.get(frameID);
                    } else {
                        sentPacket = sentPackets.remove(frameID);
                    }
                    if (sentPacket != null) {
                        increaseCount++;
                    }
                }
                increaseTokens(increaseCount);

                // deal with the current response
                SentPacketInfo sentPacket = sentPackets.get(recvFrameID);
                if (sentPacket != null) {
                    // do not increase token if have already received duplicated ack
                    if (recvFrameID > prevRecvFrameID) {    
                        increaseTokens(1);
                    }

                    if (Const.IS_EXPERIMENT) {
                        try {
                            String log = recvFrameID + "\t" + recvEngineID + "\t" +
                                    sentPacket.generatedTime + "\t" + sentPacket.compressedTime + "\t" + 
                                    receivedPacket.msgRecvTime + "\t" + receivedPacket.guidanceDoneTime + "\t" + 
                                    receivedPacket.status;
                            fileWriter.write(log + "\n");
                        } catch (IOException e) {}
                    }
                }
                prevRecvFrameID = recvFrameID;
            }
        }
    };

    public void logSentPacket(long frameID, long dataTime, long compressedTime) {
        this.sentPackets.put(frameID, new SentPacketInfo(dataTime, compressedTime));
    }

    /**
     * Blocks and only returns when token > 0
     * @return the current token number
     */
    public int getCurrentToken() {
        synchronized (tokenLock) {
            if (this.currentToken > 0) {
                return this.currentToken;
            } else {
                try {
                    tokenLock.wait();
                } catch (InterruptedException e) {}
                return this.currentToken;
            }
        }
    }

    public void increaseTokens(long count) {
        synchronized (tokenLock) {
            this.currentToken += count;
            this.tokenLock.notify();
        }
    }

    public void decreaseToken() {
        synchronized (tokenLock) {
            if (this.currentToken > 0) {
                this.currentToken--;
            }
            this.tokenLock.notify();
        }
    }

    public void close() {
        sentPackets.clear();
        if (Const.IS_EXPERIMENT) {
            try {
                fileWriter.close();
            } catch (IOException e) {
                Log.e(LOG_TAG, "Error in closing latency file");
            }
        }
    }
}

File: client/legacy-hololens-client/gabriel-holo-client/Assets/Scripts/TokenController.cs
#if !UNITY_EDITOR
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace gabriel_client
{
    class TokenController
    {
        private int _currentToken = 0;
        private System.Object _tokenLock = new System.Object();
        private ConcurrentDictionary<long, SentPacketInfo> _sentPackets = new ConcurrentDictionary<long, SentPacketInfo>();
        private long _prevRecvFrameID = 0;
        private MyLogger _myLogger;

        public TokenController(int tokenSize, MyLogger myLogger)
        {
            this._currentToken = tokenSize;
            this._myLogger = myLogger;
        }

        public int GetCurrentToken()
        {
            return _currentToken;
        }

        public void IncreaseTokens(int count)
        {
            lock (_tokenLock)
            {
                _currentToken += count;
            }
        }

        public void DecreaseToken()
        {
            lock (_tokenLock)
            {
                _currentToken--;
            }
        }

        public void LogSentPacket(long frameID, long dataTime, long compressedTime, Matrix4x4 cameraToWorldMatrix, Matrix4x4 projectionMatrix)
        {
            this._sentPackets.TryAdd(frameID, new SentPacketInfo(dataTime, compressedTime, cameraToWorldMatrix, projectionMatrix));
        }

        public SentPacketInfo GetSentPacket(long frameID)
        {
            SentPacketInfo p;
            bool isSuccess = this._sentPackets.TryGetValue(frameID, out p);
            return p;
        }

        public void ProcessReceivedPacket(ReceivedPacketInfo receivedPacket)
        {
            long recvFrameID = receivedPacket.frameID;
            string recvEngineID = receivedPacket.engineID;
            SentPacketInfo sentPacket = null;
            bool isSuccess;

            // increase appropriate amount of tokens
            int increaseCount = 0;
            for (long frameID = _prevRecvFrameID + 1; frameID < recvFrameID; frameID++)
            {
                sentPacket = null;
                if (Const.IS_EXPERIMENT)
                {
                    // Do not remove since we need to measure latency even for the late response
                    isSuccess = _sentPackets.TryGetValue(frameID, out sentPacket);
                }
                else
                {
                    isSuccess = _sentPackets.TryRemove(frameID, out sentPacket);
                }
                if (isSuccess == true)
                {
                    increaseCount++;
                }
            }
            IncreaseTokens(increaseCount);

            // deal with the current response
            isSuccess = _sentPackets.TryGetValue(recvFrameID, out sentPacket);
            if (isSuccess == true)
            {
                // do not increase token if have already received duplicated ack
                if (recvFrameID > _prevRecvFrameID)
                {
                    IncreaseTokens(1);
                }

                if (Const.IS_EXPERIMENT)
                {
                    string log = recvFrameID + "\t" + recvEngineID + "\t" +
                            sentPacket.generatedTime + "\t" + sentPacket.compressedTime + "\t" +
                            receivedPacket.msgRecvTime + "\t" + receivedPacket.guidanceDoneTime + "\t" +
                            receivedPacket.status;
                    //await _myLogger.WriteString(log + "\n");
                }
            }
            if (recvFrameID > _prevRecvFrameID)
            {
                _prevRecvFrameID = recvFrameID;
            }
        }
    }

    class SentPacketInfo
    {
        public long generatedTime;
        public long compressedTime;
        public Matrix4x4 cameraToWorldMatrix;
        public Matrix4x4 projectionMatrix;

        public SentPacketInfo(long generatedTime, long compressedTime, Matrix4x4 c, Matrix4x4 p)
        {
            this.generatedTime = generatedTime;
            this.compressedTime = compressedTime;
            this.cameraToWorldMatrix = c;
            this.projectionMatrix = p;
        }
    }

    class ReceivedPacketInfo
    {
        public long frameID;
        public string engineID;
        public string status;
        public long msgRecvTime;
        public long guidanceDoneTime;

        public ReceivedPacketInfo(long frameID, String engineID, String status)
        {
            this.frameID = frameID;
            this.engineID = engineID;
            this.status = status;
            this.msgRecvTime = -1;
            this.guidanceDoneTime = -1;
        }

        public void setMsgRecvTime(long time)
        {
            msgRecvTime = time;
        }

        public void setGuidanceDoneTime(long time)
        {
            guidanceDoneTime = time;
        }
    }
}
#endif